### DISCUSSION TITLE
Autodetect current controller type and select appropriate prompt texture

### DISCUSSION THREAD

**Discussion by Zoukambada** [emoose/OutRun2006Tweaks#224] [Suggestions]
### Autodetect the current controller input and select / swap to the appropriate community/custom prompt texture if found in textures folder

- **How it the detection would work?**

For example, if the DualShocks4 is detected (through driver or by the hardware vendor, etc) it would load the texture button prompts with corresponding to the its suffix.

1. **For dualshock 4, if detected, it would look for suffix:** SONY_DS4
First attempt: ACF61D7C_1024x512_SONY_DS4.DDS... (and other button prompt files)
Second attempt - If not found: ACF61D7C_1024x512_SONY*.dds
Third Attempt: If it finds PlayStation
Fouth Attempt: If it finds PS
Last attempt - If still not found: ACF61D7C_1024x512.dds (regular / default custom dds or the default texture would be loaded)

2. **If DualSense it is detected it would look for texture files with suffix:** SONY_DS5
First attempt:  ACF61D7C_1024x512_SONY_DS5.DDS
Second attempt: If it finds SONY_DS4 instead, as it has SONY in its filename, it would be loaded instead.
...

3. **If X360 pad is detected:**
First attempt: ACF61D7C_1024x512_MICROSOFT_XBX.DDS... (and other button prompt files)
Second attempt - ACF61D7C_1024x512_MICROSOFT_XBOX_MODERN.DDS
Third Attempt: If it finds MODERN
Third Attempt: If it finds X360
Fouth Attempt: If it finds XB*
Fifth Attempt: If it finds Microsoft*

4. **If Nintendo pad is detected:**
First attempt: ACF61D7C_1024x512_NINTENDO*
...

5. **If driving wheel is detected:**
First attempt: ACF61D7C_1024x512_WHEEL*

6. **If no controller is detected the keyboard prompt would be loaded:**
First attempt: ACF61D7C_1024x512_KEYBOARD* (regular / default custom dds or the default texture would be loaded)
...

- **Options in .INI to be set**

In OutRun2006Tweaks.ini it would be added some related options to the texture prompts / Autodetect controller feature.
**AutodetectControllerPrompts=True**

**ForceControllerPromptsSuffix=False***
*If filled with "Nintendo", "Sony" or "Microsoft" (or any other) is set, it would only search for filenames containing this suffix.

If the last option is set to anything that is not "False", add another complementary option:

**ForceControllerPromptsSuffixOverride=False**
If False, any kind of controller or wheel is detected but keyboard, the suffix texture from the last option will be loaded. When the controller is disconnected, the keyboard prompt texture is always loaded by default.
If True, the Forced Prompt Suffix will always be loaded, no matter if there is a controller connected or not.

**ForceControllerPromptsColorScheme=False****
**When "False", it would be try to load the one with suffix that has Color, if not found it would load respecting the alphabetic/numeric order of the suffix. When "Light" its the monochromatic Light scheme (light suffix), when "Dark" its forced to the monochromatic Dark scheme.

- **Textures to be detected**

If I'm not mistaken, these are the button prompt texture files to be checked:
textures/load/spr_sprani_congrats_cvt_Exst/
455717B2_512x512.dds

textures/load/spr_sprani_etc_cvt_Exst/
1F77CB88_128x128.dds

textures/load/spr_sprani_game_cvt_Exst/
39229D64_1024x1024.dds

textures/load/spr_sprani_sumo_fe_cvt_Exst/
ACF61D7C_1024x512.dds
313DB8CB_512x256.dds

- **Why?**

This way the community could share one multi-prompt package containing all updated textures, the user would download this package and place in the textures/load/FOLDER/ and it would automagically adapt to the current controller that the player is using. It would feel natural as a modern Windows / Console Game.

**Comment by Tenome** [comment]
I believe this is now a functional feature, so this one can be closed.

**Reply by Zoukambada** [reply]
Do you know how this really works?

Controller types:
PlayStation
Xbox
Switch

What path structures can be used from below:
1) textures\load\spr_sprani_congrats_cvt_Exst\Xbox\455717B2_512x512.dds
2) textures\load\Xbox\spr_sprani_congrats_cvt_Exst\455717B2_512x512.dds

What textures are detected?
455717B2, 1F77CB88, FA7BBB13, 39229D64, ACF61D7C and 31C58963 (press start / press enter)?


PS: I think it can be closed. Sometime ago me and @emoose talked about that it would also be interesting to swap sprites according to the button mapping in the new input overlay settings, instead of only be a image hard coded. I don't know if it will be related to this post or it will be another issue.
